### PR TITLE
LPD-37767 | Greater than filter for the dateModified fiels does not work properly with custom objects

### PR DESCRIPTION
**Pull request by dannielraposo**
**What's changed?**
In this PR, two key changes have been made to address issues related to date filtering for system object fields of type `DATE_TIME`:

1. **Date Format Alignment**: The query logic has been updated to ensure the correct formatting of `DATE_TIME` values when filtering `dateCreated` and `dateModified`. The changes ensure:
   - Seconds are correctly accepted and the filters work as expected.
   - Milliseconds are ignored, addressing issues where inconsistent precision (such as milliseconds) caused mismatches.

2. **Fix in `PredicateExpressionVisitorImpl._getInPredicate`**: A bug in the `_getInPredicate` method has been fixed. Previously, the right-side values in filter comparisons were not being correctly formatted, especially for dates. Since date formats vary across different databases, this led to filter failures. The PR corrects this by ensuring that dates are formatted uniformly regardless of the database being used.

Additionally, an **integration test** has been added to ensure these fixes work as intended, particularly for date filtering scenarios.

See the following Jira Ticket: [LPD-37767](https://liferay.atlassian.net/browse/LPD-37767)

### CODE REVIEW

**Comment by dannielraposo** [comment]
ci:test:relevant

**Comment by dannielraposo** [comment]
ci:stop

**Comment by dannielraposo** [comment]
ci:test:sf

**Comment by dannielraposo** [comment]
ci:test:relevant

**Comment by dannielraposo** [comment]
ci:test:relevant

**Comment by dannielraposo** [comment]
Failing tests do not seem related to PR

**Comment by dannielraposo** [comment]
Test has passed with mySQL, Hypersonic and Oracle

**Comment by dannielraposo** [comment]
ci:test:sf

**Comment by dannielraposo** [comment]
ci:test:relevant

**Comment by dannielraposo** [comment]
ci:test:sf

**Comment by dannielraposo** [comment]
ci:test:relevant

**Comment by dannielraposo** [comment]
Test passes with MySQL, PostgreSQL, Oracle and Hypersonic

**Comment by dannielraposo** [comment]
ci:stop

**Comment by dannielraposo** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/field/predicate/provider/SystemDateFieldPredicateProvider.java [hunk 1]
```diff
@@ -111,7 +111,7 @@ public Predicate getInPredicate(
 			predicate = predicate.or(binaryExpressionPredicate);
 		}
 
-		return predicate;
+		return predicate.withParentheses();
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/field/predicate/provider/SystemDateFieldPredicateProvider.java:114]: What if the predicate is null? Is this possible?

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 2]
```diff
@@ -5285,66 +5285,69 @@ public void testGetObjectEntriesFilteredByDateModified() throws Exception {
...
-			URLCodec.encodeURL("dateModified lt 2023-09-20T10:05:00.450Z"),
+			URLCodec.encodeURL("dateModified lt 2023-09-20T10:05:00Z"),
 			_objectDefinition1);
 		_assertFilterString(
 			_OBJECT_FIELD_NAME_1, _OBJECT_FIELD_VALUE_1,
-			URLCodec.encodeURL("dateModified lt 2023-09-20T10:05:00Z"),
+			URLCodec.encodeURL("dateModified lt 2023-09-20T10:05:00.999Z"),
 			_objectDefinition1);
 
 		// Test case for 'le' (less than or equal)
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:5313]: These comments must be sorted (and the use case too based on the order)

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 3]
```diff
@@ -5285,66 +5285,69 @@ public void testGetObjectEntriesFilteredByDateModified() throws Exception {
 
 		_assertFilterString(
 			_OBJECT_FIELD_NAME_1, _OBJECT_FIELD_VALUE_2,
-			URLCodec.encodeURL("dateModified gt 2023-09-20T10:00:00.150Z"),
+			URLCodec.encodeURL("dateModified gt 2023-09-20T10:00:00Z"),
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:5288]: Should we also test the dateCreated?

**Review by ccorreagg — COMMENTED** [review]
just few comments/questions

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
@dannielraposo I have included several commits for the tests. Please check them and feel free to update whatever you consider

**Review by ccorreagg — APPROVED** [review]

**Comment by dannielraposo** [comment]
ci:stop

**Comment by dannielraposo** [comment]
ci:forward

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:forward

**Pull request closed without merging**